styx-common/src/main/java/com/spotify/styx/util/ShardedCounter.java

```diff
@@ -0,0 +1,241 @@
+        datastore.runInTransaction(transaction -> {
+          final Entity shard = transaction.get(shardKey);
+          if (shard == null) {
+            transaction.put(Entity.newBuilder(shard).set(PROPERTY_SHARD, 0).build());
+          }
+          return null;
+        });
+      }
+    }
+
+    /**
+     * Gets the counter state. Creates it in Datastore if it doesn't exist yet.
+     */
+    private static CounterSnapshot fromDatastore(Datastore datastore, String counterId) {
+      CounterSnapshot snapshot = new CounterSnapshot();
+
+      final Key limitKey = datastore.newKeyFactory().setKind(KIND_COUNTER_LIMIT).newKey(counterId);
+      final Entity limit = datastore.get(limitKey);
+      if (limit == null) {
+        snapshot.limit = Long.MAX_VALUE;
+        // Or IllegalStateException("No limit found in Datastore for " + counterId);?
+      } else {
+        snapshot.limit = limit.getLong(PROPERTY_LIMIT);
+      }
+
+      final EntityQuery queryShards = EntityQuery.newEntityQueryBuilder()
+          .setKind(KIND_COUNTER_SHARD)
+          .setFilter(CompositeFilter.and(
+              PropertyFilter.ge("__key__", counterId + "-0"),
+              PropertyFilter.lt("__key__", counterId + "-" + NUM_SHARDS)))
+          .setOrderBy(OrderBy.asc("__key__"))
+          .setLimit(NUM_SHARDS)
+          .build();
+      final QueryResults<Entity> shards = datastore.run(queryShards);
+      snapshot.shards = new ArrayList<Long>(NUM_SHARDS);
+      int shardCount;
+      for (shardCount = 0; shards.hasNext(); shardCount++) {
+        Entity shard = shards.next();
+        String shardKey = shard.getKey().toString();
+        int shardIndex = Integer.valueOf(shardKey.substring(shardKey.indexOf('-') + 1));
+        snapshot.shards.set(shardIndex, shard.getLong(PROPERTY_SHARD));
+      }
+      if (shardCount < NUM_SHARDS) {
+        // The counter probably has not been initialized (so we have empty QueryResults). Also
+        // possible that a prior initialize() crashed halfway, or we got a partial list of shards in
+        // QueryResults due to eventual consistency. In any case, repeated initialization eventually
+        // creates all NUM_SHARDS shards.
+        initialize(datastore, counterId);
+      }
+
+      snapshot.updatedAt = Instant.now();
+      return snapshot;
+    }
+
+    private boolean isRecent() {
+      return updatedAt.plus(CACHE_EXPIRY_MILLIS, ChronoUnit.MILLIS).isAfter(Instant.now());
+    }
+
+    private long shardCapacity(int shardIndex) {
+      return limit / NUM_SHARDS + (shardIndex < limit % NUM_SHARDS ? 1 : 0);
+    }
+
+    /**
+     * Returns shard index which _likely_ could be successfully updated by delta, according to our
+     * cached view of the state in Datastore.
+     */
+    private int pickShardWithSpareCapacity(long delta) {
+      List<Integer> candidates = new ArrayList<>();
+
+      for (int i = 0; i < shards.size(); i++) {
+        if (shards.get(i) + delta >= 0
```

Comment:
Consider doing `Range.closed(0, shardCapacity(i)).contains(shards.get(i) + delta)`, or `shardHasCapacity(i delta)` or at least only computing `shards.get(i) + delta` once and storing it in a variable.